README.md
![Go](https://github.com/nleeper/goment/workflows/Go/badge.svg)

# Goment
###### Current Version: 1.4.2
###### [Changelog](CHANGELOG.md)

Goment is a port of the popular Javascript datetime library [Moment.js](https://momentjs.com/). It follows the Moment.js API closely, with some changes to make it more Go-like (e.g. using nanoseconds instead of milliseconds). 

Goment is still a work in progress. Please feel free to fork and contribute missing methods, locale/languages functionality, or just provide more idiomatic Go if you see some areas to improve. I have a list of things that need added/fixed in [TODO.md](TODO.md), but will create issues for them at some point.

## Features
* [Parsing](#parsing) 
* [Get+Set](#get-set)
* [Manipulate](#manipulate)
* [Display](#display)
* [Query](#query)
* [i18n](#i18n)

### Parsing
#### From now
Creates a Goment object for the current local time returned by time.Now().
```
goment.New()
```
#### From ISO 8601 string
Creates a Goment object by parsing the string as an ISO 8601 date time. The timezone will be UTC unless supplied in the string.
```
goment.New('2013-02-08 09:30:26')
```
#### From string + format
Creates a Goment object by parsing the string using the supplied format. The timezone will be the local timezone unless supplied in the string.

The parsing tokens are similar to the formatting tokens used in [Goment#Format](#format).

Goment's parser is strict, and defaults to being accurate over forgiving.

##### Year, month, and day tokens
|   | Token | Output |
| - | ----- | ------ |
| Month | M | 1 2 ... 11 12 |
| | MM | 01 01 ... 11 12 |
| | MMM | Jan Feb ... Nov Dec |
| | MMMM | January February ... November December |
| Day of Month | D | 1 2 ... 30 31 |
| | Do | 1st 2nd ... 30th 31st |
| | DD | 01 02 ... 30 31 |
| Day of Year | DDD	 | 1 2 ... 364 365 |
| | DDDD | 001 002 ... 364 365 |
| Year | YY | 70 71 ... 29 30 |
| | YYYY | 1970 1971 ... 2029 2030 |
| | Y | -25 |
| Quarter | Q | 1 2 3 4 |
| Unix Timestamp | X | 1360013296 |
| | | |
```
goment.New("12-25-1995", "MM-DD-YYYY")
```

##### Hour, minute, second, and offset tokens
|   | Token | Output |
| - | ----- | ------ |
| AM/PM	| A | AM PM |
| | a |	am pm |
| Hour| H | 0 1 ... 22 23 |
| | HH | 00 01 ... 22 23 |
| | h | 1 2 ... 11 12 |
| | hh | 01 02 ... 11 12 |
| | k | 1 2 ... 23 24 |
| | kk | 01 02 ... 23 24 |
| Minute | m | 0 1 ... 58 59 |
| | mm | 00 01 ... 58 59 |
| Second | s | 0 1 ... 58 59 |
| | ss | 00 01 ... 58 59 |
| Time Zone	| Z | -07:00 -06:00 ... +06:00 +07:00 |
| | ZZ | -0700 -0600 ... +0600 +0700 |
| | | |

##### Week year, week, and weekday tokens
For these, the lowercase tokens use the locale aware week start days, and the uppercase tokens use the ISO week date start days.

|   | Token | Output |
| - | ----- | ------ |
| Locale 4 digit week year | gggg | 2014 |
| Locale 2 digit week year | gg | 14 |
| Locale week of year | w ww | 1..53 |
| Locale day of week | e | 0..6 |
| Day name in locale set by Goment.SetLocale() | dd ddd dddd | Mo..Mon..Sunday |
| ISO 4 digit week year | GGGG | 2014 |
| ISO 2 digit week year | GG | 14 |
| ISO week of year | W WW | 1..53 |
| ISO day of week | E | 1..7 |
| | | |

#### From string + format + locale
As of Goment 1.2.0, a locale can now be supplied to parse locale-specific dates and times.

##### Locale-aware formats
|   | Token | Output |
| - | ----- | ------ |
| Time | LT | 8:30 PM |
| Time with seconds	| LTS | 8:30:25 PM |
| Month numeral, day of month, year	| L	| 09/04/1986 |
| | l | 9/4/1986 |
| Month name, day of month, year | LL | September 4, 1986 |
| | ll | Sep 4, 1986 |
| Month name, day of month, year, time | LLL | September 4, 1986 8:30 PM |
| | lll	| Sep 4, 1986 8:30 PM |
| Month name, day of month, day of week, year, time	| LLLL |	Thursday, September 4, 1986 8:30 PM |
| | llll | Thu, Sep 4, 1986 8:30 PM |
| | | |
```
goment.New("2 de septiembre de 1999 12:30", "LLL", "es")
```

#### From Unix nanoseconds
Creates a Goment object from the Unix nanoseconds since the Unix Epoch.
```
goment.New(time.Now().UnixNano())
```
#### From Unix seconds
Creates a Goment object from the Unix timestamp (seconds since the Unix Epoch).
```
goment.Unix(1318781876)
```
#### From [Go Time object](https://golang.org/pkg/time/#Time)
Creates a Goment object from the supplied Go time object.
```
goment.New(time.Date(2015, 11, 10, 5, 30, 0, 0, time.UTC))
```
#### From a Goment clone
Creates a Goment object from a clone of the supplied Goment object.
```
goment.New(goment.New('2011-05-08'))
```
#### From Goment DateTime object
Creates a Goment object from a Goment DateTime object.
``` 
goment.New(DateTime{
    Year:  2015,
    Month: 1,
    Day:   25,
    Hour:  10,
})
```

### Get+Set
#### Get
Get is a string getter using the supplied units.

##### Supported units
* y, year, years
* M, month, months
* D, date, dates
* h, hour, hours
* m, minute, minutes
* s, second, seconds
* ms, millisecond, milliseconds
* ns, nanosecond, nanoseconds

```
g.Get('hours') // 22
```
#### Nanosecond
Gets the nanoseconds of the Goment object.
```
g.Nanosecond() // 600
```
#### Millisecond
Gets the milliseconds of the Goment object.
```
g.Millisecond() // 330
```
#### Second
Gets the seconds of the Goment object.
```
g.Second() // 33
```
#### Minute
Gets the minutes of the Goment object.
```
g.Minute() // 45
```
#### Hour
Gets the hours of the Goment object.
```
g.Hour() // 22
```
#### Day of Month
Gets the day of the month of the Goment object.
```
g.Date() // 19
```
#### Day of Week
Gets the day of the week (Sunday = 0...) of the Goment object.
```
g.Day() // 2
```
#### Day of Week (Locale Aware)
Gets the day of the week according to the locale of the Goment object.
```
g.Weekday() // 3
```
#### ISO Day of Week
Gets the Goment object ISO day of the week with 1 being Monday and 7 being Sunday.
```
g.ISOWeekday() // 4
```
#### Day of Year
Gets the day of the year of the Goment object.
```
g.DayOfYear() // 100
```
#### Week of Year
Gets the localized week of the year for the Goment object.

The week of the year varies depending on which day is the first day of the week (Sunday, Monday, etc), and which week is the first week of the year.

For example, in the United States, Sunday is the first day of the week. The week with January 1st in it is the first week of the year.

In France, Monday is the first day of the week, and the week with January 4th is the first week of the year.

The output of `g.Week()` will depend on the locale for the Goment object.
```
g.Week() // 52
```
#### Week of Year (ISO)
Gets the ISO week of the year of the Goment object.
```
g.ISOWeek() // 6
```
#### Month
Gets the month (January = 1...) of the Goment object.
```
g.Month() // 2
```
#### Quarter
Gets the quarter (1 to 4) of the Goment object.
```
g.Quarter() // 1
```
#### Year
Gets the year of the Goment object.
```
g.Year() // 2013
```
#### Week Year
Gets the week-year according to the Goment object's locale.

Because the first day of the first week does not always fall on the first day of the year, sometimes the week-year will differ from the month year.

For example, in the US, the week that contains Jan 1 is always the first week. In the US, weeks also start on Sunday. If Jan 1 was a Monday, Dec 31 would belong to the same week as Jan 1, and thus the same week-year as Jan 1. Dec 30 would have a different week-year than Dec 31.
```
g.WeekYear() // 2012
```
#### Week Year (ISO)
Gets the ISO week-year of the Goment object.
```
g.ISOWeekYear() // 2013
```
#### Weeks In Year
Gets the number of weeks according to locale in the current Goment's year.
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.4.2] - 2021-04-28
### Changed
- Added cs locale support.

## [1.4.1] - 2021-04-11
### Changed
- Bugfix on `StartOf("isoWeek") not correct on Sunday #38` from kesyn

## [1.4.0] - 2020-09-04
### Added
- Added MonthByNumber method to return the month name by number.
- Added MonthShortByNumber method to return the short month name by number.
- Added WeekdayByNumber method to return the weekday name by number. This method is locale aware.

### Changed
- Updated Weekdays method to be locale aware if bool parameter is provided.
- Updated WeekdaysShort method to be locale aware if bool parameter is provided.
- Updated WeekdaysMin method to be locale aware if bool parameter is provided.
- Support for parsing timestamps with Z timezone - https://github.com/nleeper/goment/pull/36

## [1.3.0] - 2020-08-20
### Added
- Implemented Weekday & SetWeekday methods, both are locale aware.
- Implemented Week & SetWeek methods, both are locale aware.
- Implemented WeekYear & SetWeekYear methods, both are locale aware.
- Implemented WeeksInYear method, which is locale aware.
- Implemented ISOWeeksInYear method.
- Implemented SetISOWeek method.
- Implemented SetISOWeekYear method.
- Added ability to pass day name into SetDay method, day name is locale aware. 
- Added support for weekday & weekyear tokens in parsing & formatting: e, E, w, ww, W, WW, gg, GG, gggg, GGGG
- Added support for min weekday name token in parsing: dd
- Added support for StartOf("week") && EndOf("week"), both are locale aware.

### Changed
- Updated SetDay to handle values that are outside of 0-6. Values will be treated as overflow to the current value.
- Updated the constructor that takes a DateTime object to default to the current day's values when not provided for year, month & date. For example, if the constructor is called as `goment.New(DateTime{ Year: 2010 })` and today's date is `2020-08-08`, the created Goment object will be for `2010-08-08 00:00:00`.
- Updated parsing to return an error if the weekday supplied in the parsing string does not match the weekday parsed from date fields. 
```
_, err := New("Wed 08-10-2017", "ddd MM-DD-YYYY") // 8-10-2017 is a Thursday
assert.EqualError(err, "There is a mismatch between parsed weekday and expected weekday")
```

### Removed

## [1.2.0] - 2020-07-29
### Added
- Support for locales when parsing Goment datetimes. Locales now can be passed in to the `New` function like `goment.New("sábado abr 11 22:52:51 2009", "dddd MMM DD HH:mm:ss YYYY", "es")`. Refer to the `Parsing` section in the `README.md` for more information.

## [1.1.1] - 2020-07-14
### Added
- Support for `zzzz` formatting token to display timezone name, e.g. `Central Standard Time`.

## [1.1.0] - 2020-07-13
### Added
- Support for internationalization using locales when displaying Goment datetimes. Locales are currently only supported for the `Format`, `From`, `To`, `FromNow`, `ToNow` & `Calendar` methods.
- Initial supported locales are `en`, `es` and `fr`. There is a framework setup to add new languages by adding a new file in the `/locales` folder. Refer to the `i18n` section in the `README.md` for more information.
- Added `Weekdays`, `WeekdaysShort`, `WeekdaysMin`, `Months` & `MonthsShort` methods for getting locale-specific values.
- Support for `YYYYY`, `YYYYYY` & `x` datetime formats.

### Changed
- `Format` now uses functions to replace values in layout, instead of converting to a Go datetime layout and using the Time.Format method.

### Removed
- Removed internal submodules.


parse.go
package goment

import (
	"errors"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/kubax2000/goment/locales"
	"github.com/kubax2000/goment/regexps"
)

var yearIdx = 0
var monthIdx = 1
var dateIdx = 2
var hourIdx = 3
var minuteIdx = 4
var secondIdx = 5
var nanosecondIdx = 6
var weekIdx = 7
var weekdayIdx = 8

var isoDates = []isoDateFormat{
	// createISOFormat("+002006-01-02", `\+\d{6}-\d\d-\d\d`),
	// createISOFormat("-002006-01-02", `-\d{6}-\d\d-\d\d`),
	createISODateFormat("2006-01-02", `\d{4}-\d\d-\d\d`),
	createISODateFormat("2006-01", `\d{4}-\d\d`, false),
	createISODateFormat("20060102", `\d{8}`),
}

var isoTimes = []isoTimeFormat{
	createISOTimeFormat("15:04:05.9999", `\d\d:\d\d:\d\d\.\d+`),
	// createISOTimeFormat("15:04:05,9999", `\d\d:\d\d:\d\d,\d+`), // TODO - will need to replace the comma with a period in date string
	createISOTimeFormat("15:04:05", `\d\d:\d\d:\d\d`),
	createISOTimeFormat("15:04", `\d\d:\d\d`),
	createISOTimeFormat("150405.9999", `\d\d\d\d\d\d\.\d+`),
	// createISOTimeFormat("150405,9999", `\d\d\d\d\d\d,\d+`), // TODO - replace comma
	createISOTimeFormat("150405", `\d\d\d\d\d\d`),
	createISOTimeFormat("1504", `\d\d\d\d`),
	createISOTimeFormat("15", `\d\d`),
}

// var isoDates = [
//     ['YYYYYY-MM-DD', /[+-]\d{6}-\d\d-\d\d/],
//     ['YYYY-MM-DD', /\d{4}-\d\d-\d\d/],
//     ['GGGG-[W]WW-E', /\d{4}-W\d\d-\d/],
//     ['GGGG-[W]WW', /\d{4}-W\d\d/, false],
//     ['YYYY-DDD', /\d{4}-\d{3}/],
//     ['YYYY-MM', /\d{4}-\d\d/, false],
//     ['YYYYYYMMDD', /[+-]\d{10}/],
//     ['YYYYMMDD', /\d{8}/],
//     // YYYYMM is NOT allowed by the standard
//     ['GGGG[W]WWE', /\d{4}W\d{3}/],
//     ['GGGG[W]WW', /\d{4}W\d{2}/, false],
//     ['YYYYDDD', /\d{7}/]
// ]

// var isoTimes = [
//     ['HH:mm:ss.SSSS', /\d\d:\d\d:\d\d\.\d+/],
//     ['HH:mm:ss,SSSS', /\d\d:\d\d:\d\d,\d+/],
//     ['HH:mm:ss', /\d\d:\d\d:\d\d/],
//     ['HH:mm', /\d\d:\d\d/],
//     ['HHmmss.SSSS', /\d\d\d\d\d\d\.\d+/],
//     ['HHmmss,SSSS', /\d\d\d\d\d\d,\d+/],
//     ['HHmmss', /\d\d\d\d\d\d/],
//     ['HHmm', /\d\d\d\d/],
//     ['HH', /\d\d/]
// ]

type configFunc func(string, *parseConfig, locales.LocaleDetails, string)
type findTokenFunc func(string, locales.LocaleDetails) (string, string)

type isoTimeFormat struct {
	format string
	regex  *regexp.Regexp
}

type isoDateFormat struct {
	format    string
	regex     *regexp.Regexp
	allowTime bool
}

type parseConfig struct {
	isUTC             bool
	tzMinutes         int
	dayOfYear         int
	overflowDayOfYear bool
	overflowWeekday   bool
	overflowWeeks     bool
	meridiem          string
	week              map[string]int
	parsedArray       map[int]int
	date              *Goment
	locale            locales.LocaleDetails
}

type parseReplacement struct {
	configFunction    configFunc
	findTokenFunction findTokenFunc
}

var parseReplacements = map[string]parseReplacement{}

func loadParseReplacements() {
	if len(parseReplacements) > 0 {
		return
	}

	// 	// // "S":    createParseReplacement("S", "frac_second", nil, nil),
	// 	// // "SS":   createParseReplacement("SS", "frac_second", nil, nil),
	// 	// // "SSS":  createParseReplacement("SSS", "frac_second", nil, nil),

	addParseReplacement([]string{"D", "DD"}, dateIdx, regexps.MatchOneToTwo)
	addParseReplacement("Do", handleOrdinalDate, func(input string, locale locales.LocaleDetails) (string, string) {
		return findRegexString(input, locale.DayOfMonthOrdinalRegex)
	})
	addParseReplacement("DDD", handleDayOfYear, regexps.MatchOneToThree)
	addParseReplacement("DDDD", handleDayOfYear, regexps.MatchThree)

	addParseReplacement([]string{"M", "MM"}, monthIdx, regexps.MatchOneToTwo)
	addParseReplacement("MMM", handleShortMonth, func(input string, locale locales.LocaleDetails) (string, string) {
		return findRegexString(input, locale.MonthsShortRegex)
	})
	addParseReplacement("MMMM", handleLongMonth, func(input string, locale locales.LocaleDetails) (string, string) {
		return findRegexString(input, locale.MonthsRegex)
	})

	addParseReplacement("Y", handleSingleDigitYear, regexps.MatchSigned)
	addParseReplacement("YY", handleTwoDigitYear, regexps.MatchOneToTwo)
	addParseReplacement("YYYY", handleFourDigitYear, regexps.MatchOneToFour)
	addParseReplacement([]string{"YYYYY", "YYYYYY"}, yearIdx, regexps.MatchOneToSix)

	addParseReplacement("Q", handleQuarter, regexps.MatchOne)

	addParseReplacement("X", handleTimestamp, regexps.MatchTimestamp)

	addParseReplacement([]string{"h", "hh", "H", "HH"}, hourIdx, regexps.MatchOneToTwo)
	addParseReplacement([]string{"k", "kk"}, handleOneToTwentyFourTime, regexps.MatchOneToTwo)
	addParseReplacement([]string{"m", "mm"}, minuteIdx, regexps.MatchOneToTwo)
	addParseReplacement([]string{"s", "ss"}, secondIdx, regexps.MatchOneToTwo)
	addParseReplacement([]string{"a", "A"}, handleMeridiem, regexps.MatchMeridiem)

	addParseReplacement([]string{"Z", "ZZ"}, handleOffset, regexps.MatchShortOffset)

	// Week & weekday parsing
	addWeekParseReplacement("dd", handleMinDayName, func(input string, locale locales.LocaleDetails) (string, string) {
		return findRegexString(input, locale.WeekdaysMinRegex)
	})
	addWeekParseReplacement("ddd", handleShortDayName, func(input string, locale locales.LocaleDetails) (string, string) {
		return findRegexString(input, locale.WeekdaysShortRegex)
	})
	addWeekParseReplacement("dddd", handleLongDayName, func(input string, locale locales.LocaleDetails) (string, string) {
		return findRegexString(input, locale.WeekdaysRegex)
	})

	addWeekParseReplacement([]string{"d", "e", "E"}, regexps.MatchOneToTwo)

	addWeekParseReplacement([]string{"w", "ww", "W", "WW"}, handleWeekNumber, regexps.MatchOneToTwo)

	addWeekParseReplacement([]string{"gg", "GG"}, handleTwoDigitWeekYear, regexps.MatchOneToTwo)
	addWeekParseReplacement([]string{"gggg", "GGGG"}, handleWeekYear, regexps.MatchOneToFour)
	addWeekParseReplacement([]string{"ggggg", "GGGGG"}, handleWeekYear, regexps.MatchOneToSix)
}

func findRegexString(input string, rx *regexp.Regexp) (string, string) {
	found := rx.FindStringIndex(input)
	if found != nil {
		return input[found[0]:found[1]], input[found[1]:len(input)]
	}
	return "", input
}

func buildStringSlice(arg interface{}) []string {
	switch v := arg.(type) {
	case string:
		return []string{v}
	case []string:
		return v
	}
	return []string{}
}

func buildFindTokenFunction(arg interface{}) findTokenFunc {
	switch v := arg.(type) {
	case func(input string, locale locales.LocaleDetails) (string, string):
		return v
	case *regexp.Regexp:
		return func(input string, locale locales.LocaleDetails) (string, string) {
			return findRegexString(input, v)
		}
	}

	// Return a no-op findTokenFunc if the type doesn't match.
	return func(input string, locale locales.LocaleDetails) (string, string) {
		return "", input
	}
}

func addWeekParseReplacement(args ...interface{}) {
	count := len(args)

	if count < 2 {
		return
	}

	var cf configFunc
	var ftf findTokenFunc

	tokens := buildStringSlice(args[0])

	switch count {
	case 2:
		tokens = buildStringSlice(args[0])
		ftf = buildFindTokenFunction(args[1])

		cf = func(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
			config.week[token] = parseNumber(input)
		}
	case 3:
		tokens = buildStringSlice(args[0])
		cf = args[1].(func(string, *parseConfig, locales.LocaleDetails, string))
		ftf = buildFindTokenFunction(args[2])
	}

	for _, t := range tokens {
		parseReplacements[t] = parseReplacement{
			func(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
				if config.week == nil {
					config.week = map[string]int{}
				}
				cf(input, config, locale, token)
			},
			ftf,
		}
	}
}

func addParseReplacement(args ...interface{}) {
	if len(args) == 3 {
		tokens := buildStringSlice(args[0])

		var cf configFunc

		switch v := args[1].(type) {
		case func(string, *parseConfig, locales.LocaleDetails, string):
			cf = v
		case int:
			cf = func(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
				config.parsedArray[v] = parseNumber(input)
			}
		}

		ftf := buildFindTokenFunction(args[2])

		for _, t := range tokens {
			parseReplacements[t] = parseReplacement{
				cf,
				ftf,
			}
		}
	}
}

func createISOTimeFormat(format, regex string) isoTimeFormat {
	return isoTimeFormat{
		format,
		regexp.MustCompile(regex),
	}
}

func createISODateFormat(format, regex string, allowTime ...bool) isoDateFormat {
	shouldAllowTime := true
	if len(allowTime) > 0 {
		shouldAllowTime = allowTime[0]
	}

	return isoDateFormat{
		format,
		regexp.MustCompile(regex),
		shouldAllowTime,
	}
}

func parseFromFormat(date string, format string, locale locales.LocaleDetails) (time.Time, error) {
	parsedDate, err := parseToGoment(date, format, locale)
	if err != nil {
		return time.Time{}, err
	}

	return parsedDate.ToTime(), nil
}

func parseISOString(date string) (time.Time, error) {
	match := regexps.ExtendedISORegex.FindStringSubmatch(date)
	if match == nil {
		match = regexps.BasicISORegex.FindStringSubmatch(date)
	}

	allowTime := true
	dateFormat, timeFormat, tzFormat := "", "", ""

	if len(match) == 5 {
		// match[0] = matched
		// match[1] = date
		// match[2] = T
		// match[3] = time
		// match[4] = timezone

		for _, i := range isoDates {
			if i.regex.MatchString(match[1]) {
				dateFormat = i.format
				allowTime = i.allowTime
				break
			}
		}

		if dateFormat == "" {
			return time.Time{}, errors.New("No matching date format found")
		}

		if match[3] != "" {
			for _, i := range isoTimes {
				if i.regex.MatchString(match[3]) {
					separator := match[2]
					if separator == "" {
						separator = " "
					}

					timeFormat = separator + i.format
					break
				}
			}

			if timeFormat == "" {
				return time.Time{}, errors.New("No matching time format found")
			}
		}

		if !allowTime && timeFormat != "" {
			return time.Time{}, errors.New("Time part not allowed")
		}

		if match[4] != "" {
			timezoneMatch := regexps.TimeZoneRegex.FindString(match[4])
			if timezoneMatch == "Z" {
				tzFormat = "Z"
			} else if timezoneMatch != "" {
				tzFormat = "-0700"
			} else {
				return time.Time{}, errors.New("Invalid timezone format")
			}
		}

		finalFormat := dateFormat + timeFormat + tzFormat
		return time.Parse(finalFormat, date)
	}

	return time.Time{}, errors.New("Not a matching ISO-8601 date")
}

func parseToGoment(date, format string, locale locales.LocaleDetails) (*Goment, error) {
	format = expandLocaleFormats(format, locale)

	bracketMatch := regexps.BracketRegex.FindAllStringIndex(format, -1)
	bracketsFound := len(bracketMatch) > 0

	// Remove bracketed fields before parsing for tokens.
	if bracketsFound {
		for i := range bracketMatch {
			format = format[0:bracketMatch[i][0]] + format[bracketMatch[i][1]:len(format)]
		}
	}

	var match [][]string
	if match = regexps.TokenRegex.FindAllStringSubmatch(format, -1); match == nil {
		return nil, errors.New("No matches found in format")
	}

	config := &parseConfig{
		false,
		-99999,
		-1,
		false,
		false,
		false,
		"",
		nil,
		map[int]int{},
		nil,
		locale,
	}

	var found = ""
	var remaining = date

	for i := range match {
		token := match[i][0]

		if rep, ok := parseReplacements[token]; ok {
			// Find the input value matching the token.
			found, remaining = rep.findTokenFunction(remaining, locale)

			if found != "" {
				rep.configFunction(found, config, locale, token)
			}
		}
	}

	return buildFromParseConfig(config)
}

func buildFromParseConfig(config *parseConfig) (*Goment, error) {
	// Update the config values based on the meridiem.
	fixForMeridiem(config)

	// If we were able to build a complete date from the format (X), return now.
	if config.date != nil {
		return config.date, nil
	}

	// Get the current date's values.
	currentDate := currentDateArray(config)

	if config.week != nil && !keyExists(dateIdx, config.parsedArray) && !keyExists(monthIdx, config.parsedArray) {
		dayOfYearFromWeekInfo(config)
	}

	// If the date hasn't been set by a parsing function, update config values if needed.
	if config.dayOfYear > -1 {
		yearToUse := defaults(config.parsedArray, currentDate, yearIdx)

		if config.dayOfYear > daysInYear(yearToUse) || config.dayOfYear == 0 {
			config.overflowDayOfYear = true
		}

		date := time.Date(yearToUse, 1, config.dayOfYear, 0, 0, 0, 0, time.UTC)
		config.parsedArray[monthIdx] = int(date.Month())
		config.parsedArray[dateIdx] = date.Day()
	}

	// Default to current date.
	// If no year, month, day of month are given, default to today.
	// If day of month is given, default month and year.
	// If month is given, default only year.
	// If year is given, don't default anything.
	for i := 0; i < 3 && !keyExists(i, config.parsedArray); i++ {
		if _, exist := config.parsedArray[i]; !exist {
			config.parsedArray[i] = currentDate[i]
		}
	}

	// Zero out any values that weren't default, including time.
	// If its month or day, set the value to 1.
	for i := 0; i < 7; i++ {
		defaultVal := 0
		if !keyExists(i, config.parsedArray) {
			if i == 1 || i == 2 {
				defaultVal = 1
			}
			config.parsedArray[i] = defaultVal
		}
	}

	createDateFromConfig(config)

	expectedWeekday := config.date.Day()
	if config.week != nil && skeyExists("d", config.week) && config.week["d"] != expectedWeekday {
		return nil, errors.New("There is a mismatch between parsed weekday and expected weekday")
	}

	return config.date, nil
}

func dayOfYearFromWeekInfo(config *parseConfig) {
	var weekday, dow, doy, week, wy int
	var currWeek weekYear

	weekdayOverflow := false

	local, _ := New()

	w := config.week

	if skeyExists("GG", w) || skeyExists("W", w) || skeyExists("E", w) {
		dow = 1
		doy = 4

		currWeek = weekOfYear(local, dow, doy)

		wy = getDefaultWeekYear("GG", config, currWeek)

		week = getDefaultWeek("W", w, 1)
		weekday = getDefaultWeekday("E", w, 1)

		if weekday < 1 || weekday > 7 {
			weekdayOverflow = true
		}
	} else {
		dow = config.locale.Week.Dow
		doy = config.locale.Week.Doy

		currWeek = weekOfYear(local, dow, doy)

		wy = getDefaultWeekYear("gg", config, currWeek)

		week = getDefaultWeek("w", w, currWeek.week)

		if skeyExists("d", w) {
			weekday = w["d"]
			if weekday < 0 || weekday > 6 {
				weekdayOverflow = true
			}
		} else if skeyExists("e", w) {
			weekday = w["e"] + dow
			if w["e"] < 0 || w["e"] > 6 {
				weekdayOverflow = true
			}
		} else {
			weekday = dow
		}
	}

	if week < 1 || week > weeksInYear(wy, dow, doy) {
		config.overflowWeeks = true
	} else if weekdayOverflow {
		config.overflowWeekday = true
	} else {
		temp := dayOfYearFromWeeks(wy, week, weekday, dow, doy)
		config.parsedArray[yearIdx] = temp.year
		config.dayOfYear = temp.dayOfYear
	}
}

func getDefaultWeekday(token string, week map[string]int, currWeekday int) int {
	if skeyExists(token, week) {
		return week[token]
	}
	return currWeekday
}

func getDefaultWeek(token string, week map[string]int, currWeek int) int {
	if skeyExists(token, week) {
		return week[token]
	}
	return currWeek
}

func getDefaultWeekYear(token string, config *parseConfig, currWeek weekYear) int {
	if skeyExists(token, config.week) {
		return config.week[token]
	}
	if keyExists(yearIdx, config.parsedArray) {
		return config.parsedArray[yearIdx]
	}
	return currWeek.year
}

func fixForMeridiem(config *parseConfig) {
	if config.meridiem == "" {
		return
	}

	hour := config.parsedArray[hourIdx]
	isPM := strings.ToLower(config.meridiem)[0] == 'p'

	if isPM && hour < 12 {
		hour += 12
	}
	if !isPM && hour == 12 {
		hour = 0
	}

	config.parsedArray[hourIdx] = hour
}

func skeyExists(key string, m map[string]int) bool {
	_, exist := m[key]
	return exist
}
func keyExists(key int, m map[int]int) bool {
	_, exist := m[key]
	return exist
}

func createDateFromConfig(config *parseConfig) {
	loc := time.Local
	if config.isUTC {
		loc = time.UTC
	}

	datetime := DateTime{
		config.parsedArray[yearIdx],
		config.parsedArray[monthIdx],
		config.parsedArray[dateIdx],
		config.parsedArray[hourIdx],
		config.parsedArray[minuteIdx],
		config.parsedArray[secondIdx],
		config.parsedArray[nanosecondIdx],
		loc,
	}

	d, _ := New(datetime)
	if config.tzMinutes != -99999 {
		// Call this method rather than SetMinute, since it checks that minute is between 0-59.
		d.subtractMinutes(config.tzMinutes)

		// Set the offset.
		d.SetUTCOffset(config.tzMinutes)
	}

	config.date = d
}

func currentDateArray(config *parseConfig) map[int]int {
	newDate, _ := New()
	if config.isUTC {
		newDate.UTC()
	}
	return map[int]int{0: newDate.Year(), 1: newDate.Month(), 2: newDate.Date()}
}

func defaults(parsed, current map[int]int, idx int) int {
	if value, exist := parsed[idx]; exist {
		return value
	}
	if value, exist := current[idx]; exist {
		return value
	}
	return 0
}

func handleSingleDigitYear(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	y, _ := strconv.Atoi(input)
	config.parsedArray[yearIdx] = y
}

func handleOrdinalDate(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.parsedArray[dateIdx] = parseNumber(regexps.MatchOneToTwo.FindAllString(input, -1)[0])
}

func handleDayOfYear(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.dayOfYear = parseNumber(strings.TrimLeft(input, "0"))
}

func handleQuarter(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.parsedArray[monthIdx] = parseNumber(input) * 3
}

func handleTimestamp(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	ts, _ := strconv.ParseInt(input, 10, 64)
	g, _ := Unix(ts)

	config.date = g
}

func handleOneToTwentyFourTime(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.parsedArray[hourIdx] = parseNumber(input) - 1
}

func handleMeridiem(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.meridiem = input
}

func handleOffset(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.isUTC = true

	match := regexps.MatchShortOffset.FindAllString(input, -1)
	if match != nil {
		offset := match[0]
		parts := regexps.ChunkOffset.FindAllString(offset, -1)

		minutes := 0
		if len(parts) == 3 {
			minutes = (parseNumber(parts[1]) * 60) + parseNumber(parts[2])
		}
		if minutes > 0 {
			if parts[0] == "-" {
				minutes *= -1
			}
		}

		config.tzMinutes = minutes
	}
}

func handleLongMonth(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.parsedArray[monthIdx] = locale.GetMonthNumber(input)
}

func handleShortMonth(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.parsedArray[monthIdx] = locale.GetMonthShortNumber(input)
}

func handleLongDayName(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	val := locale.GetWeekdayNumber(input)
	if val != -1 {
		config.week["d"] = val
	}
}

func handleShortDayName(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	val := locale.GetWeekdayShortNumber(input)
	if val != -1 {
		config.week["d"] = val
	}
}

func handleMinDayName(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	val := locale.GetWeekdayMinNumber(input)
	if val != -1 {
		config.week["d"] = val
	}
}

func handleWeekNumber(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.week[token[:1]] = parseNumber(input)
}

func handleWeekYear(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.week[token[:2]] = parseNumber(input)
}

func handleTwoDigitWeekYear(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.week[token] = parseTwoDigitYear(input)
}

func handleTwoDigitYear(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	config.parsedArray[yearIdx] = parseTwoDigitYear(input)
}

func handleFourDigitYear(input string, config *parseConfig, locale locales.LocaleDetails, token string) {
	year := 0
	if len(input) == 2 {
		year = parseTwoDigitYear(input)
	} else {
		year = parseNumber(input)
	}
	config.parsedArray[yearIdx] = year
}

func parseTwoDigitYear(input string) int {
	baseYear := 0
	year := parseNumber(input)

	if year > 68 {
		baseYear = 1900
	} else {
		baseYear = 2000
	}

	return year + baseYear
}

func parseNumber(dateText string) int {
	num, _ := strconv.Atoi(dateText)
	return num
}
